File: src/lib/pricing/lmsr.ts
/**
 * Маркет-мейкер Хэнсона (LMSR). Порт services/robot/lmsr.py один-в-один.
 *
 *   C(q) = b · ln( e^(q_yes/b) + e^(q_no/b) )
 *   цена ДА = 1 / (1 + e^(−(q_yes − q_no)/b))
 *
 * Смысл функции стоимости появляется в РАЗНОСТИ: цена покупки = C(после) − C(до).
 * Поэтому крупная заявка стоит дороже, чем «цена × объём», — цена растёт по
 * ходу исполнения.
 */

/** Цену держим в коридоре: ровно 0 или 1 = робот берёт риск бесплатно. */
export const PRICE_FLOOR = 0.001;
export const PRICE_CEIL = 0.999;

export type LmsrOutcome = "YES" | "NO";

/** Порог, за которым exp() уходит в переполнение или в ноль. */
const EXP_LIMIT = 700;

export class LMSR {
  /** Глубина/риск: макс. убыток ограничен b · ln 2. */
  b: number;
  /** Всего куплено билетов ДА. */
  qYes: number;
  /** Всего куплено билетов НЕТ. */
  qNo: number;

  constructor(b: number, qYes = 0, qNo = 0) {
    this.b = b;
    this.qYes = qYes;
    this.qNo = qNo;
  }

  /**
   * «Сколько денег в системе» при текущих счётчиках.
   * Устойчивая версия: выносим максимум за логарифм, иначе e^x взрывается на
   * больших q. Тождество: ln(e^a + e^c) = m + ln(e^(a−m) + e^(c−m)), m = max(a, c).
   */
  cost(): number {
    const a1 = this.qYes / this.b;
    const a2 = this.qNo / this.b;
    const m = Math.max(a1, a2);
    return this.b * (m + Math.log(Math.exp(a1 - m) + Math.exp(a2 - m)));
  }

  /** Мгновенная цена ДА, зажатая в коридор. */
  priceYes(): number {
    const net = (this.qYes - this.qNo) / this.b;
    let p: number;
    if (net > EXP_LIMIT) p = 1;
    else if (net < -EXP_LIMIT) p = 0;
    else p = 1 / (1 + Math.exp(-net));
    return Math.min(Math.max(p, PRICE_FLOOR), PRICE_CEIL);
  }

  /** Цена НЕТ. По построению p_yes + p_no = 1 — тождество ДА+НЕТ = $1. */
  priceNo(): number {
    return 1 - this.priceYes();
  }

  /** Честная стоимость покупки qty билетов: C(после) − C(до). */
  buyCost(outcome: LmsrOutcome, qty: number): number {
    const before = this.cost();
    let after: number;
    if (outcome === "YES") after = new LMSR(this.b, this.qYes + qty, this.qNo).cost();
    else if (outcome === "NO") after = new LMSR(this.b, this.qYes, this.qNo + qty).cost();
    else throw new Error("outcome must be 'YES' or 'NO'");
    return after - before;
  }

  /** Зафиксировать покупку: сдвинуть счётчики, цены пересчитаются сами. */
  apply(outcome: LmsrOutcome, qty: number): void {
    if (outcome === "YES") this.qYes += qty;
    else if (outcome === "NO") this.qNo += qty;
    else throw new Error("outcome must be 'YES' or 'NO'");
  }

  /**
   * Верхняя граница убытка робота за всё время жизни рынка: b · ln 2.
   * Не зависит от того, как торговали, — только от b. Это сумма, которую
   * платформа осознанно кладёт на раскрутку рынка.
   */
  maxLoss(): number {
    return this.b * Math.LN2;
  }
}

File: backend/services/robot/lmsr.py
from __future__ import annotations

import math
from dataclasses import dataclass

PRICE_FLOOR = 0.001  # цену держим в коридоре [0.001, 0.999]:
PRICE_CEIL = 0.999   # цена ровно 0 или 1 = робот берёт риск бесплатно
@dataclass

class LMSR:
    b: float          # глубина/риск: макс. убыток ограничен b * ln 2
    q_yes: float = 0  # всего куплено билетов ДА
    q_no: float = 0   # всего куплено билетов НЕТ

    def cost(self) -> float:
        """
        Функция стоимости C(q) = b * ln( e^(q_yes/b) + e^(q_no/b) ).

        Это «сколько денег в системе» при текущих счётчиках. Сама по себе
        абстрактна; смысл появляется в РАЗНОСТИ: цена покупки = C(после) - C(до).

        Устойчивая версия: выносим максимум за логарифм, чтобы e^x не взорвалась
        на больших q. Математически это тождество:
            ln(e^a + e^b) = m + ln(e^(a-m) + e^(b-m)),  m = max(a, b)
        """
        a1 = self.q_yes / self.b
        a2 = self.q_no / self.b
        m = max(a1, a2)
        return self.b * (m + math.log(math.exp(a1 - m) + math.exp(a2 - m)))

    def price_yes(self) -> float:
        """Мгновенная цена ДА, зажатая в коридор [PRICE_FLOOR, PRICE_CEIL]."""
        net = (self.q_yes - self.q_no) / self.b
        if net > 700:       # e^(-net) ниже float -> цена -> 1
            p = 1.0
        elif net < -700:    # e^(-net) переполнил бы float -> цена -> 0
            p = 0.0
        else:
            p = 1.0 / (1.0 + math.exp(-net))
        return min(max(p, PRICE_FLOOR), PRICE_CEIL)

    def price_no(self) -> float:
        """Цена НЕТ. По построению p_yes + p_no = 1 (наше тождество YES+NO=$1)."""
        return 1.0 - self.price_yes()

    def buy_cost(self, outcome: str, qty: float) -> float:
        """
        Сколько робот возьмёт за покупку `qty` билетов исхода `outcome`.

        Ключевая идея LMSR: цена не фиксирована, а «интегрируется» по мере покупки —
        каждый следующий билет чуть дороже предыдущего. Поэтому берём не price*qty,
        а честную разность функции стоимости: C(после покупки) - C(до).
        """
        before = self.cost()
        if outcome == "YES":
            after = LMSR(self.b, self.q_yes + qty, self.q_no).cost()
        elif outcome == "NO":
            after = LMSR(self.b, self.q_yes, self.q_no + qty).cost()
        else:
            raise ValueError("outcome must be 'YES' or 'NO'")
        return after - before

    def apply(self, outcome: str, qty: float) -> None:
        """Зафиксировать покупку: сдвинуть счётчики. После — цены пересчитаются сами."""
        if outcome == "YES":
            self.q_yes += qty
        elif outcome == "NO":
            self.q_no += qty
        else:
            raise ValueError("outcome must be 'YES' or 'NO'")

    def max_loss(self) -> float:
        """
        Верхняя граница убытка робота за всё время жизни рынка: b * ln 2.
        Не зависит от того, как торговали — только от b. Это сумма, которую
        платформа осознанно кладёт на раскрутку рынка.
        """
        return self.b * math.log(2)